// @okta/courage/src/require.config.js
/* jshint -W079:false */
/* eslint no-undef: 0 */
/*exported require */
/*global okta */

var require = {
  paths: {
    okta: 'shared/util/Okta',
    jquery: 'vendor/lib/jquery-1.11.3',
    jqueryui: 'vendor/lib/jquery-ui-1.10.4.custom',
    jquerytree: 'vendor/plugins/jquery.fancytree',
    underscore: 'vendor/lib/underscore-1.8.3',
    backbone: 'vendor/lib/backbone-1.2.1',
    handlebars: 'vendor/lib/handlebars-v2.0.0',
    i18n: 'vendor/plugins/i18n',
    imagesloaded: 'vendor/plugins/imagesloaded',
    moment: 'vendor/lib/moment-with-locales',
    enduserlib: 'enduser/lib',
    qtip: 'vendor/plugins/jquery.qtip',
    'eventie/eventie': 'vendor/plugins/eventie',
    'eventEmitter/EventEmitter': 'vendor/plugins/EventEmitter',
    mixpanel: 'vendor/lib/mixpanel-2.8.0.min',
    clipboard: 'vendor/lib/clipboard-1.5.16.min'
  },
  map: {
    '*': {
      'jquery': 'vendor/lib/jquery-wrapper',
      'underscore': 'vendor/lib/underscore-wrapper',
      'jqueryui': 'vendor/lib/jqueryui-wrapper',
      'handlebars': 'vendor/lib/handlebars-wrapper',
      'moment': 'vendor/lib/moment-wrapper'
    },
    'vendor/lib/jqueryui-wrapper': {
      'jqueryui': 'jqueryui'
    },
    'vendor/lib/jquery-wrapper': {
      'jquery': 'jquery'
    },
    // due to _.template conflict
    'shared/util/StringUtil': {
      'underscore': 'underscore'
    },
    'shared/util/markdownToHtml': {
      'underscore': 'underscore',
      'handlebars': 'handlebars'
    },
    'vendor/lib/underscore-wrapper': {
      'underscore': 'underscore'
    },
    'vendor/lib/handlebars-wrapper': {
      'handlebars': 'handlebars',
      // due to _.template conflict
      'underscore': 'underscore'
    },
    'vendor/lib/moment-wrapper': {
      'moment': 'moment'
    }
  },
  shim: {
    mixpanel: {
      deps: ['vendor/lib/mixpanel-snippet'],
      exports: 'mixpanel'
    }
  },
  config: {
    i18n: {
      locale: typeof okta != 'undefined' && (okta.locale || 'en')
    }
  }
};
